File: LeetCode/Trie/JavaScript/trie.js
// A trie (pronounced as "try") or prefix tree is a tree data structure used to efficiently store and retrieve keys in a dataset of strings. There are various applications of this data structure, such as autocomplete and spellchecker.

// Implement the Trie class:

// Trie() Initializes the trie object.
// void insert(String word) Inserts the string word into the trie.
// boolean search(String word) Returns true if the string word is in the trie (i.e., was inserted before), and false otherwise.
// boolean startsWith(String prefix) Returns true if there is a previously inserted string word that has the prefix prefix, and false otherwise.

// Example 1:

// Input
// ["Trie", "insert", "search", "search", "startsWith", "insert", "search"]
// [[], ["apple"], ["apple"], ["app"], ["app"], ["app"], ["app"]]
// Output
// [null, null, true, false, true, null, true]

// Explanation
// Trie trie = new Trie();
// trie.insert("apple");
// trie.search("apple");   // return True
// trie.search("app");     // return False
// trie.startsWith("app"); // return True
// trie.insert("app");
// trie.search("app");     // return True

// Constraints:

// 1 <= word.length, prefix.length <= 2000
// word and prefix consist only of lowercase English letters.
// At most 3 * 104 calls in total will be made to insert, search, and startsWith.

// =============================================================================
// This is similar to a BST instead, we use a map, key/val pair.
// For insert method, we check the chars of the word if it exists in our map.
// TC: O(n) for insert because at worst, we'll create a chain of trie node, one for each chars of the word
// SC: O(n) for insert because at worst, we'll create a chain of trie node, one for each chars of the word
// For search and startsWith method, we check the chars of the word if it exists in our map.
// TC: O(n) for insert because at worst, we all the trie nodes until we hit the endOfWord indicator
// SC: O(1) in place processing

class TrieNode {
  constructor() {
    this.children = {}; // This is an JS Object, Key/Value Pair
    this.endOfWord = false; // This need to be a boolean
  }
}

var Trie = function () {
  // This function will create a root variable by instantiating a TrieNode()
  // This is kind off the constructor of the Trie DS
  this.root = new TrieNode();
};

/**
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert = function (word) {
  // The insert method will check of each chars in the word is in the childrens map.
  // If not, we create a new trie node for that char. We do this until we check all chars and then we mark the endOfWord to true
  let current = this.root;

  for (const c of word) {
    if (!current.children[c]) {
      current.children[c] = new TrieNode();
    }
    current = current.children[c];
  }
  current.endOfWord = true;
};

/**
 * @param {string} word
 * @return {boolean}
 */
Trie.prototype.search = function (word) {
  // The search method will check each chars in word if it exists in our children nodes.
  // If it does not exist, we return false. If we reach the end of the trie, we return the endOfWord boolean
  let current = this.root;

  for (const c of word) {
    if (!current.children[c]) {
      return false;
    }
    current = current.children[c];
  }
  return current.endOfWord;
};

/**
 * @param {string} prefix
 * @return {boolean}
 */
Trie.prototype.startsWith = function (prefix) {
  // The search method will check each chars in prefix if it exists in our children nodes.
  // If it does not exist, we return false. If we reach the end of the trie, we return true
  let current = this.root;

  for (const c of prefix) {
    if (!current.children[c]) {
      return false;
    }
    current = current.children[c];
  }
  return true;
};

/**
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */

File: LeetCode/Trie/java/Trie/Trie.java
package LeetCode.Trie.java.Trie;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

enum testResult {
    PASS("\u001B[32mPASS\u001B[0m"),
    FAIL("\u001B[31mFAIL\u001B[0m");

    private final String value;

    testResult(String value) {
        this.value = value;
    }

    public String getValue() {
        return this.value;
    }

}

record TrieRecord(String[] operations, String[][] words, List<String> expected) {
}

class TrieNode {
    private Map<Character, TrieNode> children = new HashMap<>();
    private boolean endOfWord = false;

    public TrieNode() {
    }

    public boolean existKeyInChildren(char key) {
        return this.children.containsKey(key);
    }

    public TrieNode getChildren(char key) {
        return this.children.get(key);
    }

    public void setChildren(char key, TrieNode node) {
        // this.children.compute(key, (k, v) -> (v == null) ? new TrieNode() : node);
        // this.children.computeIfAbsent(key, (k) -> new TrieNode());
        this.children.put(key, node);
    }

    public boolean getEndOfWord() {
        return this.endOfWord;
    }

    public void setEndOfWord(boolean value) {
        this.endOfWord = value;
    }
}

class Trie {

    private TrieNode root = null;

    public Trie() {
        this.root = new TrieNode();
    }

    public void insert(String word) {
        TrieNode current = this.root;

        // for each chars in the word, check all the children if the c exists in the
        // key. If not, create a K/V.
        // We do this until we check all chars and then we mark the endOfWord to true
        for (char c : word.toCharArray()) {
            if (!current.existKeyInChildren(c)) {
                current.setChildren(c, new TrieNode());
            }
            // current.initChildrenIfKeyIsAbsent(c);
            current = current.getChildren(c);
        }

        current.setEndOfWord(true);
    }

    public boolean search(String word) {
        TrieNode current = this.root;
        // The search method will check each chars in word if it exists in our children
        // nodes.
        // If it does not exist, we return false. If we reach the end of the trie, we
        // return the endOfWord boolean
        for (char c : word.toCharArray()) {
            if (!current.existKeyInChildren(c)) {
                return false;
            }
            current = current.getChildren(c);
        }

        return current.getEndOfWord();
    }

    public boolean startsWith(String prefix) {
        TrieNode current = this.root;
        // The startsWith method will check each chars in prefix if it exists in our
        // children nodes.
        // If it does not exist, we return false. If we reach the end of the trie, we
        // return true
        for (char c : prefix.toCharArray()) {
            if (!current.existKeyInChildren(c)) {
                return false;
            }
            current = current.getChildren(c);
        }
        return true;
    }

    public static void main(String[] args) {
        TrieRecord[] records = new TrieRecord[] {
                new TrieRecord(
                        new String[] { "Trie", "insert", "search", "search", "startsWith", "insert", "search" },
                        new String[][] { {}, { "apple" }, { "apple" }, { "app" }, { "app" }, { "app" }, { "app" } },
                        List.of("null", "null", "true", "false", "true", "null", "true"))
        };

        int i = 1;
        for (TrieRecord record : records) {
            System.out.println("# Test case " + i++);
            Trie.testSolution(record);
            System.out.println("-".repeat(50));
        }
    }

    private static void testSolution(TrieRecord record) {
        System.out.println("input:\toperations: " + Arrays.toString(record.operations()));
        System.out.println("\twords: " + Arrays.deepToString(record.words()));
        System.out.println("expected: " + record.expected());

        List<String> output = new ArrayList<>();
        Trie obj = null;

        for (int i = 0; i < record.operations().length; i++) {
            String operation = record.operations()[i].strip();

            switch (operation) {
                case "Trie" -> {
                    obj = new Trie();
                    output.add("null");
                }
                case "insert" -> {
                    obj.insert(record.words()[i][0]);
                    output.add("null");
                }
                case "search" -> {
                    output.add(String.valueOf(obj.search(record.words()[i][0])));
                }
                case "startsWith" -> {
                    output.add(String.valueOf(obj.startsWith(record.words()[i][0])));
                }
            }
        }

        System.out.println("result: " + output);
        System.out.println(output.equals(record.expected()) ? testResult.PASS.getValue() : testResult.FAIL.getValue());
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */